File: Data Structures & Algorithms/rotting-fruit/submission-9.py
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        visited = set()
        queue = []
        m, n = len(grid), len(grid[0])
        fresh, time = 0, 0
        for i in range(m):
            for j in range(n):
                if grid[i][j] == 1:
                    fresh += 1
                elif grid[i][j] == 2:
                    visited.add((i, j))
                    queue.append((i, j))
        if fresh == 0:
            return 0
        while queue:
            for _ in range(len(queue)):
                i, j = queue.pop(0)
                grid[i][j] = 2
                for x, y in [(0, 1), (0, -1), (1, 0), (-1, 0)]:
                    r, c = i + x, j + y
                    if (
                        not m > r >= 0 or\
                        not n > c >= 0 or\
                        grid[r][c] != 1 or\
                        (r, c) in visited
                    ): continue
                    fresh -= 1
                    queue.append((r, c))
                    visited.add((r, c))
            time += 1
        return time - 1 if fresh == 0 else -1

File: Data Structures & Algorithms/network-delay-time/submission-9.cpp
class Solution {
public:
    int networkDelayTime(vector<vector<int>>& times, int n, int k) {
        vector<vector<pair<int, int>>> adj(n + 1, vector<pair<int, int>>());
        vector<bool> visited(n + 1, false);
        vector<int> dist(n + 1, INT_MAX);
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
        int time = 0;
        dist[k] = 0;
        pq.push({0, k});
        for (const auto& t : times) {
            adj[t[0]].push_back({t[1], t[2]});
        }
        while (!pq.empty()) {
            auto curr = pq.top();
            pq.pop();
            int u = curr.second;
            if (visited[u]) {continue;}
            time = max(dist[u], time);
            for (const auto& nei : adj[u]) {
                int v = nei.first, w = nei.second;
                if (dist[u] + w < dist[v]) {
                    dist[v] = dist[u] + w;
                    pq.push({dist[v], v});
                }
            }
        }
        for (int i = 1; i <= n; i++) {
            if (dist[i] == INT_MAX) {
                return -1;
            }
        }
        return time;
    }
};
